nossomercadolivre-api/src/main/java/com/github/nossomercadolivre/validation/UniqueUserEmailValidator.java

```diff
@@ -0,0 +1,28 @@
+package com.github.nossomercadolivre.validation;
+
+import com.github.nossomercadolivre.UserRepository;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import static java.util.Objects.isNull;
+
+public class UniqueUserEmailValidator implements ConstraintValidator<UniqueUserEmail, String> {
+
+    private UserRepository userRepository;
+
+    public UniqueUserEmailValidator(UserRepository userRepository) {
```

Comment:
A injeção rolou de boa aqui? Massa! Super chique!

Reply:
Isso mesmo, o Spring varre e quando encontra um ConstraintValidator ele já entende como se fosse um @Component

O ruim, é que a annotation @UniqueUserEmail só funciona no DTO, não sei explicar porque não funciona em classes com @Entity, por isso tem um método privado isValid, na minha experiência, eventualmente acontece de precisar dessa validação em outros pontos (mesmo que não devesse, é outro caso provocado pela força do prazo de entrega), então, tendo a criar aquela sobrecarga de método, para que eu possa usar em outros lugares, OK,  isso soa como gold plating :D eu sei

Reply:
Debuga a execução, com a annotation nos dois lugares, apesar que faz mais sentido ficar no DTO mesmo. Digo, um usuário tem as restrições dele... A restrição de email duplicado é do sistema. De toda forma seria legal debugar... Meu chute é que no DTO o Spring puxa a instanciação do validator para ele(apesar que não lembro de ter visto isso quando olhei o código) e quando está no @Entity, momento onde o Hibernate se integra com a Bean Validation, é o próprio Hibernate que controla. E dessa forma ele não executa a injeção de dependência para você.